File: 2024/ocaml/bin/main.ml
module DayMap = Map.Make (Int)

module type Day = sig
  val solve : unit -> unit
  val solve_bonus : unit -> unit
end

let days =
  [
    (1, (module Solutions.Day01 : Day));
    (2, (module Solutions.Day02 : Day));
    (3, (module Solutions.Day03 : Day));
    (4, (module Solutions.Day04 : Day));
  ]
  |> List.to_seq |> DayMap.of_seq

let usage_msg = "aoc [-bonus] -d <day>"
let bonus = ref false
let day_num = ref 0

let anon_fun arg =
  Printf.fprintf stderr "Error: Unexpected argument '%s'\n" arg;
  exit 1

let () =
  let speclist =
    [
      ("-bonus", Arg.Set bonus, "Check solve_bonus function");
      ("-d", Arg.Set_int day_num, "Set the day number to check");
    ]
  in
  Arg.parse speclist anon_fun usage_msg;
  match DayMap.find_opt !day_num days with
  | Some day_module ->
      let module D = (val day_module : Day) in
      if !bonus then D.solve_bonus () else D.solve ()
  | None -> Printf.printf "Day %d not implemented yet.\n" !day_num

File: 2024/rust/src/main.rs
mod solutions;
mod utils;

use clap::Parser;
use utils::get_input;

#[derive(Parser, Debug)]
#[command(version, about, long_about=None)]
struct Args {
    #[arg(short = 'd', long = "day")]
    day: u8,

    #[arg(short = 'b', long = "bonus", default_value_t = false)]
    bonus: bool,
}

fn main() {
    let args = Args::parse();
    match args.day {
        1 => {
            if args.bonus {
                solutions::day01::bonus(&get_input(1).unwrap());
            } else {
                solutions::day01::solve(&get_input(1).unwrap());
            }
        }
        2 => {
            if args.bonus {
                solutions::day02::bonus(&get_input(2).unwrap());
            } else {
                solutions::day02::solve(&get_input(2).unwrap());
            }
        }
        3 => {
            if args.bonus {
                solutions::day03::bonus(&get_input(3).unwrap());
            } else {
                solutions::day03::solve(&get_input(3).unwrap());
            }
        }
        4 => {
            if args.bonus {
                solutions::day04::bonus(&get_input(4).unwrap());
            } else {
                solutions::day04::solve(&get_input(4).unwrap());
            }
        }
        _ => println!("Day {} is not yet implemented", args.day),
    }
}
